src/internal.c

```diff
@@ -7445,36 +7472,111 @@ int SendKexDhReply(WOLFSSH* ssh)
+                if (pubKey == NULL || privKey == NULL) {
+                    ret = WS_MEMORY_E;
+                }
+            #else
+                ecc_key pubKey[1];
+                ecc_key privKey[1];
+            #endif
 
                 if (ret == 0) {
-                    kem = OQS_KEM_new(OQS_KEM_alg_saber_lightsaber);
+                    kem = OQS_KEM_new(OQS_KEM_alg_kyber_512);
                     if (kem == NULL) {
                         ret = WS_INVALID_ALGO_ID;
                     }
                 }
 
                 if (ret == 0) {
-                    if (ssh->handshake->eSz != kem->length_public_key) {
-                        ret = WS_PUBKEY_REJECTED_E;
-                    }
+                    primeId = wcPrimeForId(ssh->handshake->kexId);
+                    if (primeId == ECC_CURVE_INVALID)
+                        ret = WS_INVALID_PRIME_CURVE;
+                }
+
+                if (ret == 0)
+                    ret = wc_ecc_init_ex(pubKey, heap, INVALID_DEVID);
+                if (ret == 0)
+                    ret = wc_ecc_init_ex(privKey, heap, INVALID_DEVID);
+#ifdef HAVE_WC_ECC_SET_RNG
+                if (ret == 0)
+                    ret = wc_ecc_set_rng(privKey, ssh->rng);
+#endif
+
+                if (ret == 0)
+                    ret = wc_ecc_import_x963_ex(ssh->handshake->e,
+                              ssh->handshake->eSz -
+                              (word32)kem->length_public_key,
+                              pubKey, primeId);
+
+                if (ret == 0)
+                    ret = wc_ecc_make_key_ex(ssh->rng,
+                              wc_ecc_get_curve_size_from_id(primeId),
+                              privKey, primeId);
+                if (ret == 0) {
+                #ifdef PRIVATE_KEY_UNLOCK
+                    PRIVATE_KEY_UNLOCK();
+                #endif
+                    ret = wc_ecc_export_x963(privKey, f_ptr, &fSz);
+                #ifdef PRIVATE_KEY_LOCK
+                    PRIVATE_KEY_LOCK();
+                #endif
+                }
+                if (ret == 0) {
+                #ifdef PRIVATE_KEY_UNLOCK
+                    PRIVATE_KEY_UNLOCK();
+                #endif
+                    ret = wc_ecc_shared_secret(privKey, pubKey,
+                                               ssh->k, &ssh->kSz);
+                #ifdef PRIVATE_KEY_LOCK
+                    PRIVATE_KEY_LOCK();
+                #endif
                 }
+                wc_ecc_free(privKey);
+                wc_ecc_free(pubKey);
+            #ifdef WOLFSSH_SMALL_STACK
+                WFREE(pubKey, heap, DYNTYPE_PUBKEY);
+                WFREE(privKey, heap, DYNTYPE_PRIVKEY);
+                pubKey  = NULL;
+                privKey = NULL;
+            #endif
 
                 if (ret == 0) {
-                    if (OQS_KEM_encaps(kem, f_ptr, ssh->k, ssh->handshake->e)
+                    if (OQS_KEM_encaps(kem, f_ptr + fSz, ssh->k + ssh->kSz,
```

Comment:
Are you checking `kem->length_public_key` < `ssh->handshake->eSz`?

Reply:
on line 7486 on new commit:
```
                if (ssh->handshake->eSz <= (word32)kem->length_public_key) {
                    ret = WS_BUFFER_E;
                }
```